File: asura-ui/src/main/scala/asura/ui/karate/KarateScenarioContext.java
package asura.ui.karate;

import java.util.function.Function;

import com.intuit.karate.CallContext;
import com.intuit.karate.LogAppender;
import com.intuit.karate.ScriptBindings;
import com.intuit.karate.core.FeatureContext;
import com.intuit.karate.core.Scenario;
import com.intuit.karate.core.ScenarioContext;

public class KarateScenarioContext extends ScenarioContext {

  public KarateScenarioContext(
    FeatureContext featureContext,
    CallContext call,
    Scenario scenario,
    LogAppender appender
  ) {
    super(featureContext, call, scenario, appender);
    doMore(this);
  }

  public KarateScenarioContext(
    FeatureContext featureContext,
    CallContext call,
    ClassLoader classLoader,
    Scenario scenario,
    LogAppender appender
  ) {
    super(featureContext, call, classLoader, scenario, appender);
    doMore(this);
  }

  @Override
  public ScenarioContext copy() {
    ScenarioContext copy = super.copy();
    doMore(copy);
    return copy;
  }

  public static void doMore(ScenarioContext ctx) {
    ctx.bindings.adds.put(ScriptBindings.READ, readX(ctx));
  }

  // TODO: support read from other storage
  public static Function<String, Object> readX(ScenarioContext ctx) {
    return s -> {
      return ctx.read.apply(s);
    };
  }

}

File: asura-ui/src/main/scala/asura/ui/karate/KarateScenarioContext.scala
package asura.ui.karate

import com.intuit.karate.core.{FeatureContext, Scenario, ScenarioContext}
import com.intuit.karate.{CallContext, LogAppender, ScriptBindings}

case class KarateScenarioContext(
                                  _featureContext: FeatureContext,
                                  _call: CallContext,
                                  _classLoader: ClassLoader,
                                  _scenario: Scenario,
                                  _appender: LogAppender,
                                  extension: KarateExtension,
                                ) extends ScenarioContext(_featureContext, _call, _classLoader, _scenario, _appender) {

  KarateScenarioContext.doMore(this)

  override def copy(): ScenarioContext = {
    val copy = super.copy
    KarateScenarioContext.doMore(copy)
    copy
  }

  override def driver(expression: String): Unit = {
    super.driver(expression)
  }
}

object KarateScenarioContext {

  def doMore(ctx: ScenarioContext): Unit = {
    if (ctx.isInstanceOf[KarateScenarioContext]) {
      val extension = ctx.asInstanceOf[KarateScenarioContext].extension
      if (extension != null && extension.driver != null) {
        ctx.setDriver(extension.driver)
      }
    }
    ctx.bindings.adds.put(ScriptBindings.READ, readX(ctx))
  }

  // TODO: support read from other storage
  def readX(ctx: ScenarioContext): String => Object = (s: String) => {
    ctx.read.apply(s)
  }
}
